gitbook-docs/request_response.md

```diff
@@ -0,0 +1,130 @@
+# Request/Response
+
+## WebSockets and similar transports
+
+The exact format of the message for a specific request is defined elsewhere in the specification.
+
+A request should include the `context` when appropriate and must include a client generated `requestId`. The `requestId` is a string and it must be a [v4 UUID](https://tools.ietf.org/html/rfc4122.html#section-4.4). It will always be included in any response to the request by the server.
+
+
+For example. a request to PUT a value:
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "put": {
+    "path": "electrical.switches.anchorLight.state",
+    "value": 1
+  }
+}
+```
+
+The server will respond with a message which includes the `requestId`, `correlationId` if provided, and a `state`.
+
+The `state` can be `PENDING` or `COMPLETED`
+
+When the state is `COMPLETED`, the message will contain a `result` value. The `result` will be any standard HTTP code including the following.
+
+- 200 - the request was succesfull
+- 502 - something went wrong carrying out the request on the server side
+- 400 - something is wrong with the client's request
+- 504 - timeout on the server side trying to carry out the request
+- 405 - the server does not support the request
+- 401 - the request has not been applied because it lacks valid authentication credentials
+- 403 - the client does not have permission to make the request
+
+The message can optionally contain a `message`.
+
+The response object may contain other objects depending on the specific request being made. For example, a response to auth request could contain a `login` object.
+
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "state": "COMPLETED",
+  "result": 200,
+  "login": {
+       "token": "....." 
+   }
+}
+```
+
+A server may respond to a request multiple times depending on how it processes the request.
+
+When a server cannot process the request immediately, it will respond with the `state` PENDING:
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "state": "PENDING"
+}
+```
+
+When processing is done, but it was not succesfull:
+```json
+{
+  "context": "vessels.self",
+  "requestId": "123345-23232-232323",
+  "state": "COMPLETED",
+  "result": 502,
```

Comment:
Result, here and in the other places, as HTTP code, makes sense to me. The key name `result` does not however. `code` is an oft-used convention when sending errors (e.g. `code: 401, message: invalid credentials`)

Reply:
How about `resultCode`, I don't really like just `code` because it's not verbose enough, IMHO.

Reply:
I guess both our arguments are 80% preferential; let's see if anyone else has an opinion here and if not we'll stick with what you have.

Reply:
How about `statusCode` since that's what they're called in the [RFC](https://tools.ietf.org/html/rfc7231#section-6) which defines them.

Reply:
I'm fine with that.